### PR TITLE
Use `Firebase Hosting Preview Channels` to deploy a web app for every pull request.

### LINKED ISSUE #8: Deploy on every PR a new Firebase Hosting App (Preview Channels)
## Description
To test and try out a PR, we should create and deploy a separate Firebase Hosting App for every PR. There is also a tutorial in the Firebase Docs, how you can do it: https://firebase.google.com/docs/hosting/github-integration?hl=en

### PR DESCRIPTION
**Pull request by nilsreichardt**
We are building for every PR a web preview, which will be deployed to Firebase Hosting. The link to the website will posted as comment (like: https://github.com/SharezoneApp/sharezone-app/pull/119#issuecomment-1030012299).

The previews are helping reviewer and other users to quickly view the changes in a compiled version.
 
A link to a preview expires after 30 days.

---

Closes #8

### CODE REVIEW

**Comment by nilsreichardt** [comment]
@Jonas-Sander Web previews are not working 👍  You can review now. They are using the `sharezone-debug` Firebase project. Could this be a problem, when users are trying to log in into their account, but the credentials are not valid for `sharezone-debug`?

**Comment by Jonas-Sander** [comment]
I think we should also add a warning in the future that its a temporary web app and that no personal data should be entered - like #27

**Comment by Jonas-Sander** [comment]
Maybe we should also add this to our cli tool `sz deploy web-app --stage preview` in the future.   
We don't have to do it in this PR, we can run with this for now.

**Comment by Jonas-Sander** [comment]
> Isn't a web app redeployed everytime a change is made? I think we should then choose at the very maximum 3 days.

Or is it 30 days from first deploy? Anyways I would lower it anyways.

**Comment by nilsreichardt** [comment]
> > Isn't a web app redeployed everytime a change is made? I think we should then choose at the very maximum 3 days.
> 
> Or is it 30 days from first deploy? Anyways I would lower it anyways.

The default is 7 days. Should we pick this or even lower, like 3 days?

**Comment by nilsreichardt** [comment]
> I think we should also add a warning in the future that its a temporary web app and that no personal data should be entered - like #27

Should we do this in this PR?

**Comment by Jonas-Sander** [comment]
> > > Isn't a web app redeployed everytime a change is made? I think we should then choose at the very maximum 3 days.
> > 
> > Or is it 30 days from first deploy? Anyways I would lower it anyways.
> 
> The default is 7 days. Should we pick this or even lower, like 3 days?

I mean there are two cases:
- the web app expires x days after the most recent deploy (so a new commit "refreshes" the expiry date)
- the web app expires x days after the first commit (so new commits don't "refresh" the expiry date)

In the first case I would say three days in the second one maybe seven (could be even more). We could also say in the latter case that we don't deploy for draft PRs so we don't start deploying a web app when it's not necessary so we don't already start the expiry date counter. 

I hope I didn't make it too complicated :D

**Comment by Jonas-Sander** [comment]
Right now it says 

```
Visit the preview URL for this PR (updated for commit 8339a78):

https://sharezone-test--pr119-web-preview-channels-q1asu2sl.web.app

(expires Sun, 06 Mar 2022 15:58:37 GMT)
```

06.03.2022 - what happens to the date if you push a new commit?

**Comment by Jonas-Sander** [comment]
> > I think we should also add a warning in the future that its a temporary web app and that no personal data should be entered - like #27
> 
> Should we do this in this PR?

No, in another one I'd say.

**Comment by nilsreichardt** [comment]
> I mean there are two cases:
> 
> * the web app expires x days after the most recent deploy (so a new commit "refreshes" the expiry date)
> * the web app expires x days after the first commit (so new commits don't "refresh" the expiry date)
> 
> In the first case I would say three days in the second one maybe seven (could be even more). We could also say in the latter case that we don't deploy for draft PRs so we don't start deploying a web app when it's not necessary so we don't already start the expiry date counter.
> 
> I hope I didn't make it too complicated :D

It expires x days after the last commit. I would not pick three days, because this would also require that we review the pull request in these three days, otherwise we are not able to use the preview. Because of this, I would pick 7 days, which gives us a bit more freedom.

**Comment by Jonas-Sander** [comment]
> It expires x days after the last commit. I would not pick three days, because this would also require that we review the pull request in these three days, otherwise we are not able to use the preview. Because of this, I would pick 7 days, which gives us a bit more freedom.

Yeah, true, let's do 7:)

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -74,3 +74,50 @@ jobs:
...
+        run: flutter build web --release
+
+      - name: Deploy to Firebase Hosting (sharezone-debug)
+        uses: FirebaseExtended/action-hosting-deploy@v0
+        with:
+          repoToken: '${{ secrets.GITHUB_TOKEN }}'
+          firebaseServiceAccount: '${{ secrets.FIREBASE_SERVICE_ACCOUNT_SHAREZONE_DEBUG }}'
+          projectId: sharezone-debug
+          entryPoint: "./app"
+          expires: '30d'
```
Comment (nilsreichardt) [inline .github/workflows/main.yml:122]:
```suggestion
          expires: '7d'
```

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -74,3 +74,50 @@ jobs:
...
         run: sz test -v
+  
+  # We are building for every PR a web preview, which will be deployed to
+  # Firebase Hosting. The link to the website will posted as comment (like:
+  # https://github.com/SharezoneApp/sharezone-app/pull/119#issuecomment-1030012299).
+  #
+  # The previews are helping reviewer and other users to quickly view the
+  # changes in a compiled version.
+  #
+  # A link to a preview expires after 30 days.
```
Comment (nilsreichardt) [inline .github/workflows/main.yml:85]:
```suggestion
  # A link to a preview expires after 7 days.
```

**Comment by nilsreichardt** [comment]
@Jonas-Sander Changed it 7d 👍 You can now review again.

**Comment by Jonas-Sander** [comment]
> I think we can just run flutter pub get inside /app and not run sz pub get (which is probably slower)

Can you fix this?

**Review by Jonas-Sander — APPROVED** [review]
See comment above

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -74,3 +74,51 @@ jobs:
...
+      # So we can just use "sz COMMAND" instead of "dart ../path/to/script.dart ..."
+      - run: echo $(realpath ./bin) >> $GITHUB_PATH
+      
+      - name: Get Flutter and Dart packages
+        run: sz pub get
+
+      - name: Build web app
+        working-directory: app
+        run: flutter build web --release
+
```
[the remark this answers is not in the document]
Comment (nilsreichardt) [reply .github/workflows/main.yml]: I'll try it out

File: .github/workflows/main.yml [hunk 4]
```diff
@@ -74,3 +74,51 @@ jobs:
...
+      - name: Deploy to Firebase Hosting (sharezone-debug)
+        uses: FirebaseExtended/action-hosting-deploy@v0
+        with:
+          repoToken: '${{ secrets.GITHUB_TOKEN }}'
+          firebaseServiceAccount: '${{ secrets.FIREBASE_SERVICE_ACCOUNT_SHAREZONE_DEBUG }}'
+          projectId: sharezone-debug
+          entryPoint: "./app"
+          expires: '30d'
```
[the remark this answers is not in the document]
Comment (nilsreichardt) [reply .github/workflows/main.yml]: This comment is outdated, right?

File: .github/workflows/main.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (nilsreichardt) [reply .github/workflows/main.yml]: Yes, it's works. It reduces the action time from 8-9 minutes to 6 minutes 👍

File: .github/workflows/main.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (Jonas-Sander) [reply .github/workflows/main.yml]: Yes

**Comment by Jonas-Sander** [comment]
LGTM

**Pull request merged**